tools/underwater-gps.py

```diff
@@ -79,101 +190,82 @@ def format(message, now=0, lat=0, lon=0, orientation=0):
     lon = abs(lon)
 
     msg = message.format(date=now.strftime("%d%m%y"),
-                       hours=now.hour,
-                       minutes=now.minute,
-                       seconds=(now.second + now.microsecond/1000000.0),
-                       lat=lat,
-                       latmin=(lat % 1) * 60,
-                       latdir=latdir,
-                       lon=lon,
-                       lonmin=(lon % 1) * 60,
-                       londir=londir,
-                       orientation=orientation)
-    return msg +  ("%02x\r\n" % calculateNmeaChecksum(msg)).upper()
-
-
-# Old mavlink link, used in BR QGC up to 3.2.4-rev6
-master = mavutil.mavlink_connection('udpout:192.168.2.1:14400', source_system=2, source_component=1)
-
-parser = argparse.ArgumentParser(description="Driver for the Water Linked Underwater GPS system.")
-parser.add_argument('--ip', action="store", type=str, default="demo.waterlinked.com", help="remote ip to query on.")
-parser.add_argument('--port', action="store", type=str, default="80", help="remote port to query on.")
-args = parser.parse_args()
-
-# New approach, UDP port 14401 for nmea data
-qgcNmeaSocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-qgcNmeaSocket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-qgcNmeaSocket.setblocking(0)
-
-connected = False
-while not connected:
-    time.sleep(5)
-    print("scanning for Water Linked underwater GPS...")
-    connected = not system('curl ' + args.ip + ':' + args.port + '/api/v1/about/')
+                         hours=now.hour,
+                         minutes=now.minute,
+                         seconds=(now.second + now.microsecond/1000000.0),
+                         lat=floor(lat),
+                         latmin=(lat % 1) * 60,
+                         latdir=latdir,
+                         lon=floor(lon),
+                         lonmin=(lon % 1) * 60,
+                         londir=londir,
+                         orientation=orientation)
+    return msg + ("%02x\r\n" % calculateNmeaChecksum(msg)).upper()
+
+
+def setup_streamrates():
+    """
+    Setup message streams to get Orientation(VFR_HUD), Depth(VFR_HUD), and temperature(SCALED_PRESSURE2)
+    """
+    # VFR_HUD at at least 5Hz
+    while not ensure_message_frequency("VFR_HUD", 5):
+        time.sleep(2)
 
-print("Found Water Linked underwater GPS!")
+    # SCALED_PRESSURE2 at at least 1Hz
+    while not ensure_message_frequency("SCALED_PRESSURE2", 1):
+        time.sleep(2)
 
-system('screen -S mavproxy -p 0 -X stuff "param set GPS_TYPE 14^M"')
 
-sockit = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-sockit.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-sockit.setblocking(0)
-sockit.bind(('0.0.0.0', 25102))
+def wait_for_waterlinked():
+    """
+    Waits until the Underwater GPS system is available
+    Returns when it is found
+    """
+    global gpsUrl
```

Comment:
Should this be a class ?
These global variables are really making this software architecture looks like a hiding class.

Reply:
Well, this whole thing can be split into at least 3 classes. I intend to do that as we move towards companion 1.0. But right now I think it is more important that our users get the waterlinked integration back.